util/src/main/java/tech/pegasys/artemis/util/backing/view/ContainerViewImpl.java

```diff
@@ -0,0 +1,102 @@
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.artemis.util.backing.view;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import tech.pegasys.artemis.util.backing.CompositeViewWrite;
+import tech.pegasys.artemis.util.backing.ContainerViewWrite;
+import tech.pegasys.artemis.util.backing.ContainerViewWriteRef;
+import tech.pegasys.artemis.util.backing.ViewRead;
+import tech.pegasys.artemis.util.backing.ViewWrite;
+import tech.pegasys.artemis.util.backing.tree.TreeNode;
+import tech.pegasys.artemis.util.backing.type.ContainerViewType;
+
+public class ContainerViewImpl<C extends ContainerViewImpl<C>>
+    extends AbstractCompositeViewWrite<C, ViewRead> implements ContainerViewWriteRef {
+
+  private final ContainerViewType<? extends ContainerViewWrite> type;
+  private TreeNode backingNode;
+
+  public ContainerViewImpl(ContainerViewType<? extends ContainerViewWrite> type) {
+    this(type, type.getDefaultTree());
+  }
+
+  public ContainerViewImpl(
+      ContainerViewType<? extends ContainerViewWrite> type, TreeNode backingNode) {
+    this.type = type;
+    this.backingNode = backingNode;
+  }
+
+  public ContainerViewImpl(
+      ContainerViewType<? extends ContainerViewWrite> type, ViewRead... memberValues) {
+    this(type, type.getDefaultTree());
+    checkArgument(
+        memberValues.length == getType().getMaxLength(),
+        "Wrong number of member values: %s",
+        memberValues.length);
+    for (int i = 0; i < memberValues.length; i++) {
+      set(i, memberValues[i]);
+    }
+  }
+
+  @Override
+  public ContainerViewType<? extends ContainerViewWrite> getType() {
+    return type;
+  }
+
+  @Override
+  public TreeNode getBackingNode() {
+    return backingNode;
+  }
+
+  @Override
+  public ViewRead get(int index) {
+    checkIndex(index);
+    TreeNode node = backingNode.get(type.treeWidth() + index);
+    return type.getChildType(index).createFromTreeNode(node);
+  }
+
+  @Override
+  public ViewWrite getByRef(int index) {
+    ViewWrite writableCopy = get(index).createWritableCopy();
+    if (writableCopy instanceof CompositeViewWrite) {
+      ((CompositeViewWrite<?>) writableCopy).setInvalidator(viewWrite -> set(index, viewWrite));
+    }
+    return writableCopy;
+  }
+
+  @Override
+  public void set(int index, ViewRead child) {
+    checkIndex(index);
+    checkArgument(
+        child.getType().equals(type.getChildType(index)),
+        "Wrong child type at index %s. Expected: %s, was %s",
+        index,
+        type.getChildType(index),
+        child.getType());
+    backingNode = backingNode.set(type.treeWidth() + index, child.getBackingNode());
```

Comment:
Might be clearer if we add a utility for calculating the generalized index: 
```suggestion
    backingNode = backingNode.set(type.getGeneralizedIndex(index), child.getBackingNode());
```

Reply:
Your way is much cleaner 👍 
https://github.com/PegaSysEng/teku/pull/1133/commits/a3a51387434e686989fe5f34d1e5535c029afee1